packages/nodes-base/nodes/MQTT/MqttTrigger.node.ts

```diff
@@ -66,96 +54,148 @@ export class MqttTrigger implements INodeType {
+		const includeTopic = options.includeTopic !== undefined ? options.includeTopic : true;
+
 		const client = await createClient(credentials);
 
-		const parsePayload = (topic: string, payload: Buffer) => {
-			let message = payload.toString();
+		const self = this;
+		const triggerFunction = async (receivedTopic: string, message: Buffer) => {
+			let messageData: IDataObject = {};
 
-			if (options.jsonParseBody) {
-				try {
-					message = JSON.parse(message);
-				} catch (e) {}
-			}
+			try {
+				const messageString = message.toString();
+				if (parseJson) {
+					try {
+						messageData = JSON.parse(messageString);
+					} catch (error) {
+						messageData = { message: messageString };
+					}
+				} else {
+					messageData = { message: messageString };
+				}
 
-			let result: IDataObject = { message, topic };
+				if (includeTopic) {
+					messageData.topic = receivedTopic;
+				}
 
-			if (options.onlyMessage) {
-				//@ts-ignore
-				result = [message];
+				const executionData: INodeExecutionData[][] = [
+					[
+						{
+							json: messageData,
+						},
+					],
+				];
+				self.emit(executionData);
+			} catch (error) {
+				const errorData: INodeExecutionData[][] = [
+					[
+						{
+							json: { error: error.message },
+						},
+					],
+				];
+				self.emit(errorData);
 			}
-
-			return [this.helpers.returnJsonArray([result])];
 		};
 
-		const manualTriggerFunction = async () =>
-			await new Promise<void>(async (resolve) => {
-				client.once('message', (topic, payload) => {
-					this.emit(parsePayload(topic, payload));
-					resolve();
-				});
-				await client.subscribeAsync(topicsQoS);
-			});
+		client.on('message', triggerFunction);
 
-		if (this.getMode() === 'trigger') {
-			const donePromise = !options.parallelProcessing
-				? this.helpers.createDeferredPromise<IRun>()
-				: undefined;
-			client.on('message', async (topic, payload) => {
-				this.emit(parsePayload(topic, payload), undefined, donePromise);
-				await donePromise?.promise;
+		await new Promise<void>((resolve, reject) => {
+			client.subscribe(topic, { qos }, (error) => {
+				if (error) {
+					reject(error);
+					return;
+				}
+				resolve();
 			});
-			await client.subscribeAsync(topicsQoS);
-		}
+		});
+
+		// Handle client errors
+		client.on('error', (error) => {
+			const errorData: INodeExecutionData[][] = [
+				[
+					{
+						json: { error: error.message },
+					},
+				],
+			];
+			this.emit(errorData);
+		});
+
+		// Handle reconnection
+		client.on('reconnect', () => {
+			const statusData: INodeExecutionData[][] = [
+				[
+					{
+						json: { status: 'reconnecting' },
+					},
+				],
+			];
+			this.emit(statusData);
+		});
+
+		client.on('close', () => {
+			const statusData: INodeExecutionData[][] = [
+				[
+					{
+						json: { status: 'disconnected' },
+					},
+				],
+			];
+			this.emit(statusData);
+		});
 
 		async function closeFunction() {
-			await client.endAsync();
+			client.end();
```

Comment:
Does this correctly result in the client ending, I think we used this in the past and had to change it